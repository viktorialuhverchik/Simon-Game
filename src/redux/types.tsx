export const POWER_OFF = "POWER_OFF";
export const POWER_ON = "POWER_ON";
export const TOGGLE_MODE = "TOGGLE_MODE";
export const START_GAME = "START_GAME";
export const MOVE_GAME = "MOVE_GAME";
export const MOVE_USER = "MOVE_USER";
export const UPDATE_SCORE = "UPDATE_SCORE";
export const TOGGLE_AVAILABLE = "TOGGLE_AVAILABLE";