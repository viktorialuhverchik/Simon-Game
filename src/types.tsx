export interface State {
    app: {
        isOn: boolean
        isStart: boolean
        isStrictMode: boolean
    },
    game: {
        movesGame: number[]
        score: number
    }
};

export interface PropsDisplay {
    isOn: boolean
    isStart: boolean
    score: number
};

export interface PropsGame {
    isStart: boolean
    isStrictMode: boolean
    movesGame: number[]
    score: number
};

export interface PropsSettings {
    isOn: boolean
    isStrictMode: boolean
    isStart: boolean
    score: number
};

export interface IThemes {
    classical: string
    synthwave: string
};
