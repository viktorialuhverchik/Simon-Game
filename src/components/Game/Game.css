.app-game {
    width: 470px;
    height: 470px;
    border-radius: 100%;
    position: relative;
    text-align: center;
    margin: auto;
    background-color: #3a3a3a;
    box-shadow: 0px 0px 8px #3a3a3a;
}

.sector {
    position: relative;
    width: 200px;
    height: 200px;
    background-color: #3a3a3a;
    border: 12px solid #3a3a3a;
}

.sector-wrapper {
    position: relative;
    top: 10px
}
  
.top-right {
    border-top-right-radius: 100%;
}
  
.bottom-right {
    border-bottom-right-radius: 100%;
}
  
.bottom-left{
    border-bottom-left-radius: 100%;
}
  
.top-left {
    border-top-left-radius: 100%;
}

.inline {
    display: inline-block;
}

.red {
    background-color: #BD2424;
}

.red.active {
    background: radial-gradient(#ef8b8b 1%, #f75b5b 40%, #bd2424);
}

.blue {
    background-color: #2127C5;
}

.blue.active {
    background: radial-gradient(#8ba6ef 1%, #5b72f7 40%, #2127C5);
}

.green {
    background-color: #33A81F;
}

.green.active {
    background: radial-gradient(#9fef8b 1%, #5bf763 40%, #33A81F);
}

.yellow {
    background-color: #D8D12B;
}

.yellow.active {
    background: radial-gradient(#fcf59d 1%, #eee060 40%, #D8D12B);
}

.disabled {
    pointer-events: none;
}

.available {
    cursor: pointer;
}
