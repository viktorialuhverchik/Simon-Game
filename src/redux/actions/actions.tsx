import React from 'react';
import {
    POWER_ON,
    POWER_OFF,
    TOGGLE_MODE,
    START_GAME,
    MOVE_GAME,
    UPDATE_SCORE, 
    TOGGLE_AVAILABLE,
    MOVE_USER
} from "../types";

export const powerOn = () => ({
    type: POWER_ON
});

export const powerOff = (isStart: boolean, score: number, dispatch: any, isAvailableClick: boolean) => {
    if(isStart) {
        dispatch(startGame(isStart));
    };
    score = 0;
    dispatch(updateScore(score));
    dispatch ({
        type: MOVE_GAME,
        moveGame: []
    });
    return {
        type: POWER_OFF
    };
};

export const sayHello = (isOn: boolean, isStart: boolean) => (
    !isOn || isStart ? "" : <h6>Hello! <br/> Select mode and press start</h6>
);

export const toggleMode = (isStrictMode: boolean) => ({
    type: TOGGLE_MODE,
    isStrictMode: !isStrictMode
});

export const startGame = (isStart: boolean) => ({
    type: START_GAME,
    isStart: !isStart
});

export const addMoveGame = (move: any) => {
    let nextStep: number = Math.round(Math.random() * 3);
    move.push(nextStep);
    console.log(move);
    return {
        type: MOVE_GAME,
        moveGame: move
    };
};

// export const addMoveUser = (step: number) => {
//     let moveUser: any = [];
//     moveUser.push(step);
//     console.log(moveUser);
//     return {
//         type: MOVE_USER,
//         moveUser
//     };
// };

export const setAvailableClick = (isAvailableClick: boolean) => ({
    type: TOGGLE_AVAILABLE,
    isAvailableClick: !isAvailableClick
});

export const updateScore = (score: number) => ({
    type: UPDATE_SCORE,
    score
});

export const compareMove = (isStrictMode: boolean, moveGame: any, moveUser: any, score: number) => {
    return (dispatch: any) => {
        // if(moveGame.length !== moveUser.length) {
        //     console.log("Loose");
        //     if(isStrictMode) {
        //         score = 0;
        //         dispatch(updateScore(score));
        //     };
        //     return;
        // };
        for (let i = 0; i < moveGame.length; i++) {
            if(moveGame[i] !== moveUser) {
                console.log("Loose");
                if(isStrictMode) {
                    score = 0;
                    dispatch(updateScore(score));
                };
            } else {
                console.log("Win");
                score++;
                console.log(score);
                dispatch(updateScore(score));
            };
        };
    };
};
