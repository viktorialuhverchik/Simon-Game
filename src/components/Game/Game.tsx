import React, { useEffect, useState } from 'react';
import { useDispatch } from 'react-redux';
import { addMoveGame, compareMove, playAudio } from '../../redux/actions/actions';

import './Game.css'; 

const Game = ({ isStart, isStrictMode, movesGame, score }: any) => {

    const dispatch = useDispatch();

    let [isAvailableClick, setIsAvailableClick] = useState(false);
    let [isActiveRed, setIsActiveRed] = useState(false);
    let [isActiveBlue, setIsActiveBlue] = useState(false);
    let [isActiveGreen, setIsActiveGreen] = useState(false);
    let [isActiveYellow, setIsActiveYellow] = useState(false);
    let [counter, setCounter] = useState(0);  
    let [isPushedRed, setIsPushedRed] = useState(false);
    let [isPushedBlue, setIsPushedBlue] = useState(false);
    let [isPushedGreen, setIsPushedGreen] = useState(false);
    let [isPushedYellow, setIsPushedYellow] = useState(false);

    useEffect(() => {
        if(!isStart) return;
        dispatch(addMoveGame(movesGame));
        setCounter(0);
        setIsAvailableClick(false);
        let gameMoveIndex = 0;

        const showMove = () => {
            switch (movesGame[gameMoveIndex]) {
                case 0:
                    setIsActiveYellow(true);
                    playAudio("/music/0.wav");
                    setTimeout(() => setIsActiveYellow(false), 1000);
                    break;
                case 1:
                    setIsActiveRed(true);
                    playAudio("/music/1.wav");
                    setTimeout(() => setIsActiveRed(false), 1000);
                    break;
                case 2:
                    setIsActiveGreen(true);
                    playAudio("/music/2.wav");
                    setTimeout(() => setIsActiveGreen(false), 1000);
                    break;
                case 3:
                    setIsActiveBlue(true);
                    playAudio("/music/3.wav");
                    setTimeout(() => setIsActiveBlue(false), 1000);
                    break;
            };
            gameMoveIndex++;

            if (movesGame.length === gameMoveIndex) {
                setIsAvailableClick(true);
            } else {
                setTimeout(showMove, 1500);
            }
        };
        
        setTimeout(showMove, 500);
    }, [isStart, score, movesGame, dispatch]);

    const checkUserMove = (step: number) => {
        switch (step) {
            case 0:
                setIsPushedYellow(true);
                setTimeout(() => setIsPushedYellow(false), 200);
                break;
            case 1:
                setIsPushedRed(true);
                setTimeout(() => setIsPushedRed(false), 200);
                break;
            case 2:
                setIsPushedGreen(true);
                setTimeout(() => setIsPushedGreen(false), 200);
                break;
            case 3:
                setIsPushedBlue(true);
                setTimeout(() => setIsPushedBlue(false), 200);
                break;
        };
        compareMove(isStrictMode, movesGame, step, score, counter)(dispatch);
        setCounter(++counter);
    };

    return (
        <div className="app-game">
            <div className="sector-wrapper">
                <div
                    className={`sector top-left inline yellow
                        ${!isActiveYellow ? "" : "active"}
                        ${!isAvailableClick ? "disabled" : "available"}
                        ${!isPushedYellow ? "" : "pushed"}`}
                    onClick={() => checkUserMove(0)}
                />
                <div
                    className={`sector top-right inline red
                        ${!isActiveRed ? "" : "active"}
                        ${!isAvailableClick ? "disabled" : "available"}
                        ${!isPushedRed ? "" : "pushed"}`}
                    onClick={() => checkUserMove(1)}
                />
                <div
                    className={`sector bottom-left inline green
                        ${!isActiveGreen ? "" : "active"}
                        ${!isAvailableClick ? "disabled" : "available"}
                        ${!isPushedGreen ? "" : "pushed"}`}
                    onClick={() => checkUserMove(2)}
                />
                <div
                    className={`sector bottom-right inline blue
                        ${!isActiveBlue ? "" : "active"}
                        ${!isAvailableClick ? "disabled" : "available"}
                        ${!isPushedBlue ? "" : "pushed"}`}
                    onClick={() => checkUserMove(3)}
                />
            </div>
        </div>
    );
};

export default Game;
