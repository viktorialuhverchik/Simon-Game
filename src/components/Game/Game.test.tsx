import React from 'react';
import { renderWithRedux } from '../../index.test';
import Game from './Game';
import fetchMock from "jest-fetch-mock";
import { fireEvent } from '@testing-library/react';
import { addMoveGame } from '../../redux/actions/actions';
import { MOVES_GAME } from '../../redux/types';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

let isOn: boolean = true;
let isStart: boolean = true;
let isStrictMode: boolean = false;
let movesGame: any = [1, 2];
let score: number = 0;

const initialState = {
    app: {
        isOn,
        isStart,
        isStrictMode
    },
    game: {
        movesGame,
        score
    }
};

describe('Game component',() => {

    let store: any;

    beforeEach(() => {
        store = mockStore(initialState);
        fetchMock.resetMocks();
    });

    afterEach(() => {
        fetchMock.mockClear();
    });

    it('renders Game component', () => {
        const wrapper = renderWithRedux(<Game isStart={isStart} isStrictMode={isStrictMode} movesGame={movesGame} score={score} />);
        expect(wrapper).toMatchSnapshot();
    });

    // it("should render game buttons", () => {
    //     const { getByTestId } = renderWithRedux(<Game isStart={isStart} isStrictMode={isStrictMode} movesGame={movesGame} score={score} />);
    //     expect(getByTestId("yellow")).toBeTruthy();
    //     expect(getByTestId("blue")).toBeTruthy();
    //     expect(getByTestId("red")).toBeTruthy();
    //     expect(getByTestId("green")).toBeTruthy();
    // });

    it('should add move game', () => {
        let movesGame: any = [];
        const expectedActions = { type: MOVES_GAME, movesGame };
        store.dispatch(addMoveGame(movesGame));
        expect(store.getActions()[0]).toEqual(expectedActions);
    });

    it('click on yellow, red, blue, green', () => {
        const { getByTestId } = renderWithRedux(<Game isStart={isStart} isStrictMode={isStrictMode} movesGame={movesGame} score={score}/>);
        fireEvent.click(getByTestId("yellow"));
        fireEvent.click(getByTestId("red"));
        fireEvent.click(getByTestId("blue"));
        fireEvent.click(getByTestId("green"));
        expect(getByTestId("yellow")).toBeDefined();
        expect(getByTestId("red")).toBeDefined();
        expect(getByTestId("blue")).toBeDefined();
        expect(getByTestId("green")).toBeDefined();
    });

    it("setTimeout", () => {
        jest.useFakeTimers();
        let callback = jest.fn();
        setTimeout(callback, 200);
        jest.runAllTimers();
    });
});