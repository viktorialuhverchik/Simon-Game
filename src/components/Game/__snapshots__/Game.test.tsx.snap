// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game component renders Game component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="app-game"
        data-testid="game"
      >
        <div
          class="sector-wrapper"
        >
          <div
            class="sector top-left inline yellow
                        
                        disabled
                        "
            data-testid="yellow"
          />
          <div
            class="sector top-right inline red
                        
                        disabled
                        "
            data-testid="red"
          />
          <div
            class="sector bottom-left inline green
                        
                        disabled
                        "
            data-testid="green"
          />
          <div
            class="sector bottom-right inline blue
                        
                        disabled
                        "
            data-testid="blue"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="app-game"
      data-testid="game"
    >
      <div
        class="sector-wrapper"
      >
        <div
          class="sector top-left inline yellow
                        
                        disabled
                        "
          data-testid="yellow"
        />
        <div
          class="sector top-right inline red
                        
                        disabled
                        "
          data-testid="red"
        />
        <div
          class="sector bottom-left inline green
                        
                        disabled
                        "
          data-testid="green"
        />
        <div
          class="sector bottom-right inline blue
                        
                        disabled
                        "
          data-testid="blue"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
